MXE: Improve parallel build

orgads opened the pull request:
If `-j` is not used, the packages are built serially (but each
package can still be built in parallel by itself).

If `JOBS` is not used, there is a limit of 6 parallel jobs.

orgads commented:
Can you suggest a better way? From what I remember, your buildbot machine has 2 cores so this doesn't disturb you. I run the build on a machine with 24 cores, and it runs much slower than it can.

lephilousophe commented:
Two things here:
- specifying both JOBS and -j would (theorically) start n² processes at the same time. This is not wanted.
- By default jobs depend on the CPU number and are maximized by the DEFAULT_MAX_JOBS (which equals to 6). It's simpler to override this variable in settings.mk. I have a patch pending.

orgads commented:
* It's not n² (I don't believe I still remember the alt+0178 sequence for this :D), because make uses a jobserver that keeps the parallel jobs up to the number you specify.
* Fine with me. But what value will you put there?

lephilousophe commented:
I did the check and the make jobserver isn't used here because MXE calls make with -j $(JOBS) everywhere.
Granted it wasn't n² (French keyboard do have a key for this ;) ) but it was greater than n.
Any large value is fine here (like 10000) as it's a max and will never be used.

orgads commented:
So I suppose this should be fixed in mxe then.

Regarding settings.mk, does it apply also on local package builds? Because when I added plugins/gcc11, gcc 5 was still built until I added it to `local_mxe_package` definition [here](https://github.com/scummvm/dockerized-bb/pull/34/files#diff-1d6ef559e082e82c7e2674cdd48316864ac0764e7d66bbb04b0212d441e5c448R11).

lephilousophe commented:
> Regarding settings.mk, does it apply also on local package builds? Because when I added plugins/gcc11, gcc 5 was still built until I added it to `local_mxe_package` definition [here](https://github.com/scummvm/dockerized-bb/pull/34/files#diff-1d6ef559e082e82c7e2674cdd48316864ac0764e7d66bbb04b0212d441e5c448R11).

I would say it's because you didn't override. When you specify MXE_PLUGIN_DIRS on command line, standard affectation of values is ignore by GNU Make. You need the override keyword for this.

orgads commented:
Oh thanks. I'll try that.

lephilousophe commented:
I have pushed the JOBS part. Can you try and see if it's enough?

orgads commented:
It's better, but still building one package at a time can be very slow, since the configure phase is very long, and it is not parallelized.

From my experience, there aren't *many* packages that are built in parallel anyway (I've noticed 8 at most), and even if they all spawn n compilers, it's not that bad. Notice that some of the packages are actually quite small, so they don't do much anyway.

If you'd run all the packages in a single command then I'd reconsider using `-j` without a jobserver, but with the way it works currently I'm quite confident it should be ok.

orgads commented:
By "small" I mean this. Notice the package sizes:
```
[build]       mxe-conf                x86_64-w64-mingw32.shared
[build]       mxe-conf                x86_64-pc-linux-gnu
[done]        mxe-conf                x86_64-w64-mingw32.shared                            12 KiB         0m2.900s
[done]        mxe-conf                x86_64-pc-linux-gnu                                  12 KiB         0m3.083s
[build]       pkgconf                 x86_64-pc-linux-gnu
[build]       mingw-w64               x86_64-w64-mingw32.shared
[build]       zstd                    x86_64-pc-linux-gnu
[build]       binutils                x86_64-w64-mingw32.shared
[done]        mingw-w64               x86_64-w64-mingw32.shared                            16 KiB         0m2.984s
[build]       gmp                     x86_64-pc-linux-gnu
[done]        pkgconf                 x86_64-pc-linux-gnu                                  1732 KiB       0m5.322s
[build]       pkgconf                 x86_64-w64-mingw32.shared
[done]        pkgconf                 x86_64-w64-mingw32.shared                            424 KiB        0m2.852s
[done]        zstd                    x86_64-pc-linux-gnu                                  16040 KiB      0m16.031s
[done]        gmp                     x86_64-pc-linux-gnu                                  29860 KiB      0m33.377s
[build]       isl                     x86_64-pc-linux-gnu
[build]       mpfr                    x86_64-pc-linux-gnu
[done]        isl                     x86_64-pc-linux-gnu                                  42100 KiB      0m19.920s
[done]        mpfr                    x86_64-pc-linux-gnu                                  19228 KiB      0m22.411s
[build]       mpc                     x86_64-pc-linux-gnu
[done]        binutils                x86_64-w64-mingw32.shared                            580288 KiB     1m7.620s
[done]        mpc                     x86_64-pc-linux-gnu                                  6712 KiB       0m12.127s
[build]       gcc                     x86_64-w64-mingw32.shared
[done]        gcc                     x86_64-w64-mingw32.shared                            4234460 KiB    6m58.507s
[meta]        cc                      x86_64-w64-mingw32.shared
```

`binutils` and `gcc` are the only large (and long-compiling) packages, and `gcc` depends on `binutils`, so it doesn't start before it is done anyway.

lephilousophe commented on toolchains/mxe/packages/toolchain/build.sh:
```diff
@@ -23,12 +26,12 @@ cd "${MXE_DIR}"/
 # Override PREFIX on command line and not in settings.mk because else it gets overriden too late
 
-make PREFIX="${MXE_PREFIX_DIR}" check-requirements
+make $MAKEFLAGS check-requirements
```
Could you not make use of the unquoted variable and add everywhere -j$JOBS? I think it's more clear about what happens.

orgads replied:
Done

orgads commented:
Ok, I changed to `nproc` everywhere instead of the long expression. Do you see a reason not to? It exists in the base image.

lephilousophe commented:
I don't see any reason. It's used inside our chosen image so it should be safe. We don't have portability issues here.
Thanks!
